### PR TITLE
Allow to run migrations in check mode (dry run)

### PR DESCRIPTION
**Pull request by fatkodima**
I have found myself a lot in situations where it could be great to have ability to run migrations to report what changes they would have made rather than making them.

Of course, this can be done by running `rake db:migrate:status` to find targeted migrations, inspect their code for changes and mentally assemble those pieces in one flow. But this is tedious, slightly time consuming and error prone even for 3 migrations. And can be automated.

`DRY_RUN` option was added to `db:migrate`, `db:rollback`, `db:migrate:up` and `db:migrate:down`. For `db:migrate:up` and `db:migrate:down` it is not very much needed, but was added for consistency and to avoid possible surprises.

The **output** in check mode looks like this:
```
▶ rake db:migrate DRY_RUN=true
== 20171231134346 CreateActiveStorageTables: migrating ========================
-- create_table(:active_storage_blobs)
-- create_table(:active_storage_attachments)
== 20171231134346 CreateActiveStorageTables: migrated =========================

== 20180103151354 CreateUsers: migrating ======================================
-- create_table(:users)
== 20180103151354 CreateUsers: migrated =======================================

== 20180103151422 CreatePosts: migrating ======================================
-- create_table(:posts)
== 20180103151422 CreatePosts: migrated =======================================

== 20180103151456 AddEmailToUsers: migrating ==================================
-- add_column(:users, :email, :string)
== 20180103151456 AddEmailToUsers: migrated ===================================
```

If this feature will be ok for core members, I'll finish this PR by adding tests, documentation, etc.
Feedback would be greatly appreciated.

### CODE REVIEW

File: activerecord/lib/active_record/tasks/database_tasks.rb [hunk 1]
```diff
@@ -187,6 +215,10 @@ def target_version
         ENV["VERSION"].to_i if ENV["VERSION"] && !ENV["VERSION"].empty?
       end
 
+      def dry_run
+        ENV["DRY_RUN"] ? ENV["DRY_RUN"] == "true" : false
```
Comment (javierhonduco) [inline activerecord/lib/active_record/tasks/database_tasks.rb]: What do you think on just checking that something is that that environment variable is not `nil`? I think that some developers may think that `DRY_RUN=1 [...]` would work

Something like:
```ruby
def dry_run
  !!ENV["DRY_RUN"]
end
```

File: activerecord/lib/active_record/tasks/database_tasks.rb [hunk 2]
```diff
@@ -169,12 +169,40 @@ def migrate
...
           scope.blank? || scope == migration.scope
         end
         ActiveRecord::Base.clear_cache!
       ensure
         Migration.verbose = verbose_was
+        Migrator.dry_run = dry_run_was
+      end
+
+      def rollback
+        step = ENV["STEP"] ? ENV["STEP"].to_i : 1
```
Comment (javierhonduco) [inline activerecord/lib/active_record/tasks/database_tasks.rb]: Minor nit: I find it's easier to read `ENV.fetch("STEP", 1).to_i`

**Comment by javierhonduco** [comment]
Damn, just saw this line
> If this feature will be ok for core members, I'll finish this PR by adding tests, documentation, etc.
Feedback would be greatly appreciated.

**Comment by matthewd** [comment]
It seems dangerous to offer a dry-run mode when we're executing arbitrary user-supplied code: we can neuter the standard migration methods, but who knows what else someone might be doing inside their migrations. At best, it'll likely fail because preceding migrations haven't put the DB into the expected state; at worst, it could make real changes to data. :confused:

**Comment by fatkodima** [comment]
@matthewd yes, very reasonable comment!
Unfortunately, looks like we can't neuter all possible code user would like to execute inside his migrations, at least not directly related for db migrations.
Wdyt, should this be closed or maybe there is a possible solution to that problem that I'm missing?

**Comment by javierhonduco** [comment]
I think the only semi-safe way would be parsing those migration files and checking that its AST only has calls to ActiveRecord's migrations methods. This would probably not play well with other ways of running migrations that could be legit, so wouldn't work 100%, unfortunately.

**Comment by rafaelfranca** [comment]
Given the already pointed problems and the fact the need for this is only for debug proposes I prefer to not add this complexity to the framework. Thank you so much for the pull request.

**Pull request closed without merging**